### DISCUSSION TITLE
[Solved] Podman executime time is faster than baremetal?

### DISCUSSION THREAD

**Discussion by GrabbenD** [containers/podman#19441] [General]
I wanted to compare the execution speed of a (rootful) Podman container versus baremetal and was surprised with the result. I've done this by re-running the same command and not counting the first result:

Setup:
```console
# Deps
$ sudo podman run \
  --rm \
  --name=benchmark \
  --privileged \
  -it \
  archlinux:latest
$ pacman --noconfirm -Syu time git

# Test
$ /usr/bin/time pacman --noconfirm -S git
```

Results (for package reinstallation from disk cache):
**Edit: see below**
 
How come Pacman is faster than baremetal?

**Comment by eriksjolund** [comment]
Maybe because of the `--rm` flag and the use of _volatile overlay mount_?
https://www.redhat.com/sysadmin/container-volatile-overlay-mounts

**Reply by GrabbenD** [reply]
I believe my test was somehow flawed, I just re-did it with and without `--rm` and got a more realistic result both times:

```console
# Arch Linux Baremetal (bash)
0.09user 0.05system 0:00.58elapsed 26%CPU (0avgtext+0avgdata 60876maxresident)k
0.11user 0.03system 0:00.57elapsed 26%CPU (0avgtext+0avgdata 60148maxresident)k
0.08user 0.06system 0:00.57elapsed 25%CPU (0avgtext+0avgdata 61008maxresident)k
0.09user 0.05system 0:00.58elapsed 25%CPU (0avgtext+0avgdata 61760maxresident)k

# Arch Linux Rootful Podman (bash)
0.17user 0.06system 0:00.27elapsed 86%CPU (0avgtext+0avgdata 75724maxresident)k
0.16user 0.06system 0:00.27elapsed 86%CPU (0avgtext+0avgdata 76300maxresident)k
0.18user 0.05system 0:00.27elapsed 87%CPU (0avgtext+0avgdata 76492maxresident)k
0.18user 0.05system 0:00.27elapsed 86%CPU (0avgtext+0avgdata 76416maxresident)k
```

---

Here's a better test which is consistent:
```console
$ podman run \
  --rm \
  --name=benchmark \
  --privileged \
  -it \
  archlinux:latest

$ pacman --noconfirm -Syu time

$ echo '#!/bin/bash
for i in $(seq 1 1000);
do bash -c ":" ;
done' > bash_benchmark_script.sh

$ /usr/bin/time bash bash_benchmark_script.sh
```

```console
# Arch Linux Baremetal (bash)
0.48user 0.15system 0:00.62elapsed 102%CPU (0avgtext+0avgdata 3584maxresident)k
0.47user 0.14system 0:00.60elapsed 102%CPU (0avgtext+0avgdata 3584maxresident)k
0.47user 0.12system 0:00.59elapsed 102%CPU (0avgtext+0avgdata 3840maxresident)k
0.46user 0.13system 0:00.58elapsed 102%CPU (0avgtext+0avgdata 3584maxresident)k

# Arch Linux Rootful Podman (bash)
0.49user 0.16system 0:00.63elapsed 102%CPU (0avgtext+0avgdata 3584maxresident)k
0.48user 0.13system 0:00.60elapsed 101%CPU (0avgtext+0avgdata 3584maxresident)k
0.48user 0.14system 0:00.61elapsed 101%CPU (0avgtext+0avgdata 3584maxresident)k
0.48user 0.16system 0:00.62elapsed 102%CPU (0avgtext+0avgdata 3584maxresident)k
```

**Comment by afbjorklund** [comment]
Why privileged? Just makes it unrealistic, for no particular reason

**Reply by GrabbenD** [reply]
I'm using it to run a stateless gaming container, I need `--privileged` as otherwise permission is denied to `/dev/tty0` ;)
https://github.com/containers/podman/discussions/18870

**Reply by afbjorklund** [reply]
Sad that `--device` is not enough for that... Anyway, the benchmark seemed a bit artificial (why not add `git` to the image?)

**Reply by GrabbenD** [reply]
> Sad that --device is not enough for that...

I tried 4 different distros (Fedora Server, OpenSuSE Tumbleweed, Arch Linux & Alpine) and had the same result. Sure, you can't replicate this issue in Fedora Workstation in nested mode since it already has a running DE with elogind service but I'm trying to run everything in containers leaving the host as barebone as possible.

> Anyway, the benchmark seemed a bit artificial (why not add git to the image?)

Do you have a example @afbjorklund ?

**Reply by afbjorklund** [reply]
Like

FROM image
RUN add git package

To make a new image

**Reply by GrabbenD** [reply]
Well, the first test was indeed artificial since network latency can affect the result (as Pacman checks if there's a new package version before installing git package from cache):

```console
# Test speed of reinstalling git from disk cache 5 times
$ /usr/bin/time pacman --noconfirm -S git

# Arch Linux Baremetal (bash)
0.09user 0.05system 0:00.58elapsed 26%CPU (0avgtext+0avgdata 60876maxresident)k
0.11user 0.03system 0:00.57elapsed 26%CPU (0avgtext+0avgdata 60148maxresident)k
0.08user 0.06system 0:00.57elapsed 25%CPU (0avgtext+0avgdata 61008maxresident)k
0.09user 0.05system 0:00.58elapsed 25%CPU (0avgtext+0avgdata 61760maxresident)k

# Arch Linux Rootful Podman (bash)
0.17user 0.06system 0:00.27elapsed 86%CPU (0avgtext+0avgdata 75724maxresident)k
0.16user 0.06system 0:00.27elapsed 86%CPU (0avgtext+0avgdata 76300maxresident)k
0.18user 0.05system 0:00.27elapsed 87%CPU (0avgtext+0avgdata 76492maxresident)k
0.18user 0.05system 0:00.27elapsed 86%CPU (0avgtext+0avgdata 76416maxresident)k
```

I only did the test to get a general idea of the overhead. To get more realistic numbers, better tests would have to be carried out with something like Phoronix Test Suite
